src/lib.rs

```diff
@@ -227,41 +227,78 @@ pub mod pallet {
 			Ok(Default::default())
 		}
 
-		/// First claim for collecting vested tokens. This one is free
-		#[pallet::weight((0, DispatchClass::Normal, Pays::No))]
-		pub fn my_first_claim(origin: OriginFor<T>) -> DispatchResultWithPostInfo {
+		/// Collect whatever portion of your reward are currently vested. The first time each
+		/// contributor calls this function pays no fees
+		#[pallet::weight(0)]
+		pub fn show_me_the_money(origin: OriginFor<T>) -> DispatchResultWithPostInfo {
 			let payee = ensure_signed(origin)?;
-
-			// Calculate the vested amount on demand.
-			let info = AccountsPayable::<T>::get(&payee).ok_or(Error::<T>::NoAssociatedClaim)?;
-
+			let initialized = <Initialized<T>>::get();
 			ensure!(
-				info.free_claim_done == false,
-				Error::<T>::FirstClaimAlreadyDone
+				initialized == true,
+				Error::<T>::RewardVecNotFullyInitializedYet
 			);
+			// Calculate the veted amount on demand.
+			let mut info =
+				AccountsPayable::<T>::get(&payee).ok_or(Error::<T>::NoAssociatedClaim)?;
 			ensure!(
 				info.claimed_reward < info.total_reward,
 				Error::<T>::RewardsAlreadyClaimed
 			);
 
-			Self::make_vested_payment(info, payee.clone(), true)
-		}
+			// Vesting is done in relation with the relay chain slot
+			let now: T::BlockNumber =
+				cumulus_pallet_parachain_system::Pallet::<T>::validation_data()
+					.expect("validation data was set in parachain system inherent")
+					.relay_parent_number
+					.into();
 
-		/// Collect whatever portion of your reward are currently vested.
-		#[pallet::weight(0)]
-		pub fn show_me_the_money(origin: OriginFor<T>) -> DispatchResultWithPostInfo {
-			let payee = ensure_signed(origin)?;
+			// Substract the first payment from the vested amount
+			let first_paid = T::InitializationPayment::get() * info.total_reward;
 
-			// Calculate the veted amount on demand.
-			let info = AccountsPayable::<T>::get(&payee).ok_or(Error::<T>::NoAssociatedClaim)?;
-			ensure!(
-				info.claimed_reward < info.total_reward,
-				Error::<T>::RewardsAlreadyClaimed
-			);
+			// Vesting perdiods as balance
+			let vesting_period = T::VestingPeriod::get()
+				.saturated_into::<u128>()
+				.try_into()
+				.ok()
+				.ok_or(Error::<T>::WrongConversionU128ToBalance)?;
 
-			Self::make_vested_payment(info, payee.clone(), false)
-		}
+			// To calculate how much could the user have claimed already
+			let payable_period = now.saturating_sub(<InitRelayBlock<T>>::get().into());
 
+			let pay_period_as_balance: BalanceOf<T> = payable_period
+				.saturated_into::<u128>()
+				.try_into()
```

Comment:
I'm not sure I understand the need for this conversation.
To know the should_have_claimed, we need: 
payable_period (number of blocks) / vesting_period (number of blocks) * (info.total_reward (balance) - first_paid (balance))

Reply:
The main reason is that I need all these have the same type to do the arithmetic computation. And I cannot divide, then make the conversion, then multiply (the division would already be truncated). I might be able to do this in a more elegant way, let me take a closer look, but this was the most straight-forward way I found

Reply:
I agree that converting a number of blocks to a balance is a bit odd. My inclination would be to convert everything to an integer when they need a common type, but that might be my C++ background talking :)

Reply:
I tried minimizing conversions but I can do that too definitely

Reply:
I ended up using `relay_chain::BlockNumber` from `cumulus_primitives_core`. I think this is what makes most sense as we are relying on this type when we count vested blocks. This type is a u32 and convertible automatically to Balance thanks to the atLeast32BitsUnsigned trait. Let me know if this looks better